### DISCUSSION TITLE
serde: deserialize a Table to a struct

### DISCUSSION THREAD

**Discussion by JulianGmp** [toml-rs/toml#744] [Q&A] [answered]
Hi there,

I'm in the following situation: I get an already parsed `toml::Table` (aka `toml::map::Map<String, toml::value::Value>`) and I want to now "deserialize" that to a struct I define. I put serde's `Deserialize` proc_macro on my type, but I couldn't find a way to deserialize the data from anything but a string.
Is there something I missed or this is currently not possible?

I saw that [`toml_edit::de::ValueDeserializer`](https://docs.rs/toml_edit/latest/toml_edit/de/struct.ValueDeserializer.html) mentions that it
> Can be created [...] from parsed values using serde::de::IntoDeserializer::into_deserializer.

I didn't the equivalent in the `toml` crate. Is that feature specific to `toml_edit`? And if so, can I safely mix both crates in one project?

For context: I'm writing custom preprocessors for [mdbook](https://rust-lang.github.io/mdBook/for_developers/preprocessors.html). A preprocessor is simply a program that's called by mdbook which gets fed all of its data via stdin. All of the options written in the `book.toml` that apply to that preprocessor (i.e. they're in the toml section of that preprocessor) are parsed by mdbook and at the end I receive that Table.  
Right now I only have 2-3 options, so manually checking the table is not a huge problem, but it means I have to do all the error handling myself or slap `unwrap` everywhere, which I'd like to avoid.

**Comment by epage** [comment, marked as answer]
I think what you want is [Table::try_into](https://docs.rs/toml/latest/toml/type.Table.html#method.try_into)

**Reply by JulianGmp** [reply]
Ah, that does seem to be the right way.
I'm still an amateur when it comes to Rust so I'm not 100% sure why try_into expects an owned value instead of a non-mutable borrow, but oh well.

The bigger problem for my project specifically is that mdbook uses `toml` version `0.5.11` and it seems the try_into implementation came with a later version. Though it seems they do [plan to make changes](https://github.com/rust-lang/mdBook/issues/2037) to this, but that's mdbook's concern not toml.

Thanks for the quick reply :)

**Reply by JulianGmp** [reply]
Update: for anyone running into this with mdbook specifically, you can use `mdbook::config::Config::get_deserialized_opt` to get a deserialized value from the mdbook API.
This avoids having to pull in the toml crate yourself and the potential version conflict that comes with it.

It's not used in mdbook's example preprocessor, [yet](https://github.com/rust-lang/mdBook/pull/2407).
